### ISSUE TITLE
Failed to convert model to tensorrt with dynamic batch-size

### ISSUE THREAD

**Issue by PhuongNDVN** [ultralytics/yolov5#9764]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Export

### Bug

The command: 
python export.py --data openstore/dataset.yaml --weights openstore/yolov5m.pt --imgsz 384 640 --include engine --device 0 --dynamic --batch-size 20

The logs:
export: data=openstore/dataset.yaml, weights=['openstore/yolov5m.pt'], imgsz=[384, 640], batch_size=20, device=0, half=False, inplace=False, keras=False, optimize=False, int8=False, dynamic=True, simplify=False, opset=12, verbose=False, workspace=4, nms=False, agnostic_nms=False, topk_per_class=100, topk_all=100, iou_thres=0.45, conf_thres=0.25, include=['engine']
YOLOv5 🚀 v6.2-189-g2f1eb21 Python-3.8.13 torch-1.11.0 CUDA:0 (NVIDIA GeForce RTX 3060, 12045MiB)

Fusing layers... 
Model summary: 290 layers, 20877180 parameters, 0 gradients, 47.9 GFLOPs

PyTorch: starting from openstore/yolov5m.pt with output shape (20, 15120, 12) (40.3 MB)

ONNX: starting export with onnx 1.12.0...
ONNX: export success ✅ 3.2s, saved as openstore/yolov5m.onnx (79.7 MB)

TensorRT: starting export with TensorRT 8.2.5.1...
[10/11/2022-15:03:09] [TRT] [I] [MemUsageChange] Init CUDA: CPU +463, GPU +0, now: CPU 3965, GPU 3390 (MiB)
[10/11/2022-15:03:10] [TRT] [I] [MemUsageSnapshot] Begin constructing builder kernel library: CPU 3965 MiB, GPU 3390 MiB
[10/11/2022-15:03:10] [TRT] [I] [MemUsageSnapshot] End constructing builder kernel library: CPU 4119 MiB, GPU 3434 MiB
[10/11/2022-15:03:10] [TRT] [I] ----------------------------------------------------------------
[10/11/2022-15:03:10] [TRT] [I] Input filename:   openstore/yolov5m.onnx
[10/11/2022-15:03:10] [TRT] [I] ONNX IR version:  0.0.7
[10/11/2022-15:03:10] [TRT] [I] Opset version:    12
[10/11/2022-15:03:10] [TRT] [I] Producer name:    pytorch
[10/11/2022-15:03:10] [TRT] [I] Producer version: 1.11.0
[10/11/2022-15:03:10] [TRT] [I] Domain:           
[10/11/2022-15:03:10] [TRT] [I] Model version:    0
[10/11/2022-15:03:10] [TRT] [I] Doc string:       
[10/11/2022-15:03:10] [TRT] [I] ----------------------------------------------------------------
[10/11/2022-15:03:10] [TRT] [W] onnx2trt_utils.cpp:366: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[10/11/2022-15:03:10] [TRT] [E] ModelImporter.cpp:773: While parsing node number 290 [Range -> "onnx::Reshape_466"]:
[10/11/2022-15:03:10] [TRT] [E] ModelImporter.cpp:774: --- Begin node ---
[10/11/2022-15:03:10] [TRT] [E] ModelImporter.cpp:775: input: "onnx::Range_464"
input: "onnx::Range_463"
input: "onnx::Range_465"
output: "onnx::Reshape_466"
name: "Range_290"
op_type: "Range"

[10/11/2022-15:03:10] [TRT] [E] ModelImporter.cpp:776: --- End node ---
[10/11/2022-15:03:10] [TRT] [E] ModelImporter.cpp:779: ERROR: builtin_op_importers.cpp:3353 In function importRange:
[8] Assertion failed: inputs.at(0).isInt32() && "For range operator with dynamic inputs, this version of TensorRT only supports INT32!"
TensorRT: export failure ❌ 3.8s: failed to load ONNX file: openstore/yolov5m.onnx

### Environment

OS: ubuntu 20.04
Python: 3.8
Pytorch: 1.12.0+cu113, Cuda: 11.3
TensorRT 8.2.5.1

### Additional

I can convert to tensorrt with fixed batch-size normally but failed when adding the --dynamic option

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@PhuongNDVN 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. Your command works correctly for me in Google Colab, so there are no YOLOv5 issues, everything is working correctly if you have your environment set up correctly.

<img width="1275" alt="Screenshot 2022-10-13 at 01 01 58" src="https://user-images.githubusercontent.com/26833433/195463523-e875939b-771b-476d-8e5c-a4d3621bdb29.png">

We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by dinobei** [comment]
Hi,
I also have the exactly same issue in Jetson-Nano dev board.

**Comment by zhudongwork** [comment]
I also have the exactly same issue in yolov5 6.0, py3.7 torch1.10 onnx1.12

**Comment by Smart-Bobo** [comment]
I have fix this problem by change export.py

yolov5 official export.py：

`
def export_onnx(model, im, file, opset, dynamic, simplify, prefix=colorstr('ONNX:')):
    # YOLOv5 ONNX export
    check_requirements('onnx')
    import onnx

    LOGGER.info(f'\n{prefix} starting export with onnx {onnx.__version__}...')
    f = file.with_suffix('.onnx')

    output_names = ['output0', 'output1'] if isinstance(model, SegmentationModel) else ['output0']
    if dynamic:
        dynamic = {'images': {0: 'batch', 2: 'height', 3: 'width'}}  # shape(1,3,640,640)
        if isinstance(model, SegmentationModel):
            dynamic['output0'] = {0: 'batch', 1: 'anchors'}  # shape(1,25200,85)
            dynamic['output1'] = {0: 'batch', 2: 'mask_height', 3: 'mask_width'}  # shape(1,32,160,160)
        elif isinstance(model, DetectionModel):
            dynamic['output0'] = {0: 'batch', 1: 'anchors'}  # shape(1,25200,85)
`

I change this code to：

`
def export_onnx(model, im, file, opset, dynamic, simplify, prefix=colorstr('ONNX:')):
    # YOLOv5 ONNX export
    check_requirements('onnx')
    import onnx

    LOGGER.info(f'\n{prefix} starting export with onnx {onnx.__version__}...')
    f = file.with_suffix('.onnx')

    output_names = ['output0', 'output1'] if isinstance(model, SegmentationModel) else ['output0']
    if dynamic:
        dynamic = {'images': {0: 'batch'}}  # shape(1,3,640,640)
        if isinstance(model, SegmentationModel):
            dynamic['output0'] = {0: 'batch'}  # shape(1,25200,85)
            dynamic['output1'] = {0: 'batch'}  # shape(1,32,160,160)
        elif isinstance(model, DetectionModel):
            dynamic['output0'] = {0: 'batch'}  # shape(1,25200,85)
`

Then This Problem Fixed！！！

**Comment by bodong-zelos** [comment]
it is a bug with "--dynamic" using tensorrt 7.x, but no problem with tensorrt 8.x.
Even with @Smart-Bobo 's modification, the engine can be build without error, but the inference results are wrong.

**Comment by glenn-jocher** [comment]
@bodong-zelos maybe we should deprecate TRT 7 support? Is this still required for some hardware like Jetson nano?

**Comment by bodong-zelos** [comment]
> @bodong-zelos maybe we should deprecate TRT 7 support? Is this still required for some hardware like Jetson nano?

It is better to support TRT7 now, as some NVIDIA hardwares (such as xavier) are still using jetpack 4.X with tensorrt 7.x for better compatibility. Jetpack 5.x is just released, and some third party software and plugins are not well supported.

**Comment by glenn-jocher** [comment]
@bodong-zelos got it, thanks!

**Comment by tigerzjh** [comment]
@glenn-jocher what is the version yolov5 you use to convert dynamic trt. I use V6.1 and find it is quit difference between dynic and constant bs.
![image](https://user-images.githubusercontent.com/23514705/201913643-4aa1e2dc-fb5d-4508-98b4-826a00799c3b.png)

**Comment by glenn-jocher** [comment]
@tigerzjh use master for everything.

Yes dynamic will be different as it needs to compute grids for the image every forward pass.

**Comment by gubei528** [comment]
TensorRT: starting export with TensorRT 8.2.1.8...
[11/23/2022-19:43:07] [TRT] [I] [MemUsageChange] Init CUDA: CPU +448, GPU +0, now: CPU 2948, GPU 9498 (MiB)
[11/23/2022-19:43:08] [TRT] [I] [MemUsageSnapshot] Begin constructing builder kernel library: CPU 2948 MiB, GPU 9498 MiB
[11/23/2022-19:43:08] [TRT] [I] [MemUsageSnapshot] End constructing builder kernel library: CPU 3102 MiB, GPU 9542 MiB
[11/23/2022-19:43:08] [TRT] [I] ----------------------------------------------------------------
[11/23/2022-19:43:08] [TRT] [I] Input filename:   yolov5s.onnx
[11/23/2022-19:43:08] [TRT] [I] ONNX IR version:  0.0.6
[11/23/2022-19:43:08] [TRT] [I] Opset version:    13
[11/23/2022-19:43:08] [TRT] [I] Producer name:    pytorch
[11/23/2022-19:43:08] [TRT] [I] Producer version: 1.8
[11/23/2022-19:43:08] [TRT] [I] Domain:           
[11/23/2022-19:43:08] [TRT] [I] Model version:    0
[11/23/2022-19:43:08] [TRT] [I] Doc string:       
[11/23/2022-19:43:08] [TRT] [I] ----------------------------------------------------------------
[11/23/2022-19:43:08] [TRT] [W] onnx2trt_utils.cpp:366: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[11/23/2022-19:43:08] [TRT] [E] ModelImporter.cpp:773: While parsing node number 210 [Range -> "354"]:
[11/23/2022-19:43:08] [TRT] [E] ModelImporter.cpp:774: --- Begin node ---
[11/23/2022-19:43:08] [TRT] [E] ModelImporter.cpp:775: input: "352"
input: "351"
input: "353"
output: "354"
name: "Range_218"
op_type: "Range"

[11/23/2022-19:43:08] [TRT] [E] ModelImporter.cpp:776: --- End node ---
[11/23/2022-19:43:08] [TRT] [E] ModelImporter.cpp:779: ERROR: builtin_op_importers.cpp:3352 In function importRange:
[8] Assertion failed: inputs.at(0).isInt32() && "For range operator with dynamic inputs, this version of TensorRT only supports INT32!"
Have you ever encountered such a problem?

**Comment by glenn-jocher** [comment]
@gubei528 no I don't recognize the error. TRT export with --dynamic works for me in Colab:

<img width="851" alt="Screenshot 2022-11-23 at 14 22 41" src="https://user-images.githubusercontent.com/26833433/203557561-ca54ec21-f537-4151-91bf-6f134ff9c20e.png">

**Comment by gubei528** [comment]
I used the same command python export.py --weights yolov5n.pt --include engine --dynamic --device 1, but there is still an error, should I convert to onnx first, and then use this command?

**Comment by glenn-jocher** [comment]
@gubei528 no you don't need to export to onnx first, it does that automatically when exporting to engine. You might just have environment issues. You can try exporting in Docker image:

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

**Comment by gubei528** [comment]
ok, i'll see tomorrow

**Comment by glenn-jocher** [comment]
@gubei528 great! Let me know if you encounter any other issues. Have a good day!
